### PR TITLE
PETSc fixes

### PR DESCRIPTION
**Pull request by connorjward**
# Description

Updates Firedrake to accept PETSc/SLEPc 3.22.2

Related PRs:

https://github.com/firedrakeproject/petsc/pull/24, https://github.com/firedrakeproject/slepc/pull/9

~Needs https://gitlab.com/petsc/petsc/-/issues/1678~

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -48,6 +48,8 @@ jobs:
       COMPLEX: ${{ matrix.complex }}
       RDMAV_FORK_SAFE: 1
     steps:
+      - name: Install pkg-config  # just for testing, will be put into firedrake-env
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:51]: Should this be removed now?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -56,12 +58,12 @@ jobs:
           rm -rf firedrake_venv
       - name: Build Firedrake
         run: |
+          unset PETSC_DIR PETSC_ARCH SLEPC_DIR
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:61]: Why are these being unset?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -83,16 +85,20 @@ jobs:
             --install defcon \
             --install gadopt \
             --install asQ \
+            --package-branch petsc connorjward/merge-upstream \
+            --package-branch slepc connorjward/merge-upstream \
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:89]:
```suggestion
```
Once these have been merged.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -56,12 +58,12 @@ jobs:
           rm -rf firedrake_venv
       - name: Build Firedrake
         run: |
+          unset PETSC_DIR PETSC_ARCH SLEPC_DIR
           cd ..
           # Linting should ignore unquoted shell variable $COMPLEX
           # shellcheck disable=SC2086
           ./firedrake/scripts/firedrake-install \
             $COMPLEX \
-            --honour-petsc-dir \
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:64]: As above

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -83,16 +85,20 @@ jobs:
             --install defcon \
             --install gadopt \
             --install asQ \
+            --package-branch petsc connorjward/merge-upstream \
+            --package-branch slepc connorjward/merge-upstream \
             || (cat firedrake-install.log && /bin/false)
       - name: Install test dependencies
         run: |
+          unset PETSC_DIR PETSC_ARCH SLEPC_DIR
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:93]: As above

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -83,16 +85,20 @@ jobs:
...
           . ../firedrake_venv/bin/activate
           python "$(which firedrake-clean)"
           python -m pip install \
             pytest-xdist pytest-timeout ipympl
           python -m pip list
       - name: Test Firedrake
         run: |
+          unset PETSC_DIR PETSC_ARCH SLEPC_DIR
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:101]: As above

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -118,6 +124,7 @@ jobs:
       - name: Test pyadjoint
         if: ${{ matrix.scalar-type == 'real' }}
         run: |
+          unset PETSC_DIR PETSC_ARCH SLEPC_DIR
```
Comment (JHopeCollins) [inline .github/workflows/build.yml:127]: As above

**Review by JHopeCollins — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 2 above)
Comment (JHopeCollins) [reply .github/workflows/build.yml:61]: In which case, it all looks good to me. I can approve once the workflow is reverted and tests are passing.

Comment (connorjward) [reply .github/workflows/build.yml:61]: Now that I have some written approval I will probably just force things through (assuming things pass) so I don't need to bother you with extra work. Thanks!

**Comment by connorjward** [comment]
I want to run the tests one more time to make sure we haven't hit a performance regression. I have been keeping the runners busy today which might account for the slow CI time.

**Pull request merged**